File: 4/solution.py
from itertools import groupby


def solve(data: str) -> str:
    # Parse input
    range_min, range_max = [int(a) for a in data.splitlines()[0].split('-')]

    passwords_found = 0  # Part 1
    proper_passwords_found = 0  # Part 2
    for i in range(range_min, range_max):
        # Split i into digits
        digits = [int(c) for c in str(i)]

        # Group digits by similar
        digits = [(key, len(list(value))) for key, value in groupby(digits)]

        # Check digits match conditions
        pair_exists = False
        two_same = False
        decrease = False
        prev_digit = -1
        for digit, count in digits:
            if digit < prev_digit:
                decrease = True
                break
            if count >= 2:
                two_same = True
            if count == 2:
                pair_exists = True

            prev_digit = digit

        # Update counts
        if not decrease and two_same:
            passwords_found += 1
            if pair_exists:
                proper_passwords_found += 1

    return f"{passwords_found} {proper_passwords_found}"

File: 4/solution.rs
mod question4 {
    use crate::common::get_digits;

    pub fn solve(data: Vec<&str>) -> String {
        // Parse input
        let range_ints: Vec<_> = data[0].split('-').map(|x| x.parse().unwrap()).collect();
        let min = range_ints[0];
        let max = range_ints[1];

        // Part 1: Count number matching initial conditions
        let mut part1_count = 0;
        let mut part2_count = 0;
        for num in min..max {
            // Get digits
            let digits = get_digits(num);
            // Check conditions
            let mut two_same = false;
            let mut only_two_same = false;
            let mut increasing = true;
            let mut cur_same = 0;
            for (d1, d2) in digits[0..digits.len() - 1].iter().zip(digits[1..].iter()) {
                if d2 < d1 {
                    increasing = false;
                    break;
                }
                if d1 == d2 {
                    two_same = true;
                    if cur_same == 0 {
                        cur_same = 2;
                    } else {
                        cur_same += 1;
                    }
                } else {
                    if cur_same == 2 {
                        only_two_same = true;
                    }
                    cur_same = 0;
                }
            }
            if cur_same == 2 {
                only_two_same = true;
            }

            if two_same && increasing {
                part1_count += 1;
                if only_two_same {
                    part2_count += 1;
                }
            }
        }

        // Return solutions
        return format!("{} {}", part1_count, part2_count);
    }
}
